python-checks/src/main/java/org/sonar/python/checks/cdk/IamPolicyPublicAccessCheck.java

```diff
@@ -0,0 +1,161 @@
+    if (callExpression.arguments().isEmpty()) {
+      return;
+    }
+
+    getArgument(subscriptionContext, callExpression, "obj", 0)
+      .flatMap(e -> e.getExpression(e2 -> e2.is(Tree.Kind.DICTIONARY_LITERAL)))
+      .map(DictionaryLiteral.class::cast)
+      .ifPresent(dictionaryLiteral -> checkJsonEntry(subscriptionContext, dictionaryLiteral));
+  }
+
+  private static void checkPolicyDocumentJson(SubscriptionContext subscriptionContext, CallExpression callExpression) {
+    Optional<DictionaryLiteral> dictionaryLiteral = getArgument(subscriptionContext, callExpression, "obj", 0)
+      .flatMap(e -> e.getExpression(isDictionaryLiteral()))
+      .map(DictionaryLiteral.class::cast);
+
+    if (dictionaryLiteral.isEmpty()) {
+      return;
+    }
+
+    Optional<ListLiteral> statement = CdkUtils.getDictionaryPair(subscriptionContext, dictionaryLiteral.get(), "Statement")
+      .flatMap(resolvedKeyValuePair -> resolvedKeyValuePair.value.getExpression(isListLiteral()))
+      .map(ListLiteral.class::cast);
+
+    if (statement.isEmpty()) {
+      return;
+    }
+
+    for (Expression element : statement.get().elements().expressions()) {
+      getDictionary(element).ifPresent(innerDict -> checkJsonEntry(subscriptionContext, innerDict));
+    }
+  }
+
+  private static void checkJsonEntry(SubscriptionContext subscriptionContext, DictionaryLiteral dictionaryLiteral) {
+    CdkUtils.DictionaryAsMap map = CdkUtils.DictionaryAsMap.build(subscriptionContext, dictionaryLiteral);
+
+    Optional<Expression> sensitiveEffect = map.get("Effect", isString("Allow"));
+    if (sensitiveEffect.isEmpty()) {
+      return;
+    }
+
+    // Given an entry under the key "Principal", we want to raise an issue if the value is either
+    //  1) string literal "*",
+    //  2) dictionary in the format of { "AWS": "*" } or { "AWS": ["99999", ..., "*", ....] }
+    map.getFlow("Principal").ifPresent(principal -> {
+      principal.addIssueIf(isString("*"), ISSUE_MESSAGE, IssueLocation.preciseLocation(sensitiveEffect.get(), SECONDARY_MESSAGE));
+
+      principal.getExpression(expression -> expression.is(Tree.Kind.DICTIONARY_LITERAL))
+        .map(DictionaryLiteral.class::cast)
+        .flatMap(innerDict -> getDictionaryPair(subscriptionContext, innerDict, "AWS"))
+        .ifPresent(aws -> {
+          // Check if the principal value is set to { "AWS": "*" }
+          aws.value.addIssueIf(isString("*"), ISSUE_MESSAGE, IssueLocation.preciseLocation(sensitiveEffect.get(), SECONDARY_MESSAGE));
+
+          // Check if the principal value is set to { "AWS": ["99999", ..., "*", ....] }
+          getListExpression(aws.value).ifPresent(listLiteral -> listLiteral.elements().expressions().stream()
+            .filter(isString("*"))
+            .forEach(
+              sensitivePrincipalElement -> raiseIssue(subscriptionContext, sensitivePrincipalElement, sensitiveEffect.get())));
```

Comment:
Why don't you make use of the flow `addIssue` method?

Reply:
I was having trouble with putting the issue highlighting on the asterisk with the `addIssue` methods (because the expression flow is a level higher):

```
  ["*", "9999999"]
#^^^
```
vs.
```
  ["*", "9999999"]
#^^^^^^^^^^^^^^
```

I will try to experiment with a few helpers if I can get the precise location of the list element.